src/numerics/petsc_matrix.C

```diff
@@ -925,6 +925,81 @@ void PetscMatrix<T>::_get_submatrix(SparseMatrix<T> & submatrix,
   petsc_submatrix->close();
 }
 
+template <typename T>
+void PetscMatrix<T>::_get_submatrix_nosort(SparseMatrix<T> & submatrix,
+                                           const std::vector<numeric_index_type> & rows,
+                                           const std::vector<numeric_index_type> & cols) const
+{
+  if (!this->closed())
+    {
+      libmesh_deprecated();
+      libmesh_warning("The matrix must be assembled before calling PetscMatrix::create_submatrix_nosort().\n"
+                      "Please update your code, as this warning will become an error in a future release.");
+      const_cast<PetscMatrix<T> *>(this)->close();
+    }
+
+  // Make sure the SparseMatrix passed in is really a PetscMatrix
+  PetscMatrix<T> * petsc_submatrix = cast_ptr<PetscMatrix<T> *>(&submatrix);
+
+  PetscErrorCode ierr=0;
+
+  ierr=MatZeroEntries(petsc_submatrix->_mat);
+  LIBMESH_CHKERR(ierr);
+
+  PetscInt pc_ncols = 0;
+  const PetscInt * pc_cols;
+  const PetscScalar * pc_vals;
+
+  // // data for creating the submatrix
+  std::vector<PetscInt> sub_cols;
+  std::vector<PetscScalar> sub_vals;
+
+  for (auto i : index_range(rows))
+  {
+    PetscInt sub_rid[] = {static_cast<PetscInt>(i)};
+    PetscInt rid = static_cast<PetscInt>(rows[i]);
+    // only get value from local rows, and set values to the corresponding columns in the submatrix
+    if (rows[i]>= this->row_start() && rows[i]< this->row_stop())
+    {
+      // get one row of data from the original matrix
+      ierr = MatGetRow(_mat, rid, &pc_ncols, &pc_cols, &pc_vals);
+      LIBMESH_CHKERR(ierr);
+      // extract data from certain cols, save the indices and entries sub_cols and sub_vals
+      for (auto j : index_range(cols))
+      {
+        for (unsigned int idx = 0; idx< static_cast<unsigned int>(pc_ncols); idx++)
+        {
+          if (pc_cols[idx] == static_cast<PetscInt>(cols[j]))
+            {
+              sub_cols.push_back(static_cast<PetscInt>(j));
+              sub_vals.push_back(pc_vals[idx]);
+            }
+        }
+      }
+      // set values
+      ierr = MatSetValues(petsc_submatrix->_mat,
+                          1,
+                          sub_rid,
+                          static_cast<PetscInt>(sub_vals.size()),
+                          sub_cols.data(),
+                          sub_vals.data(),
+                          INSERT_VALUES);
+      LIBMESH_CHKERR(ierr);
+      ierr = MatRestoreRow(_mat, rid, &pc_ncols, &pc_cols, &pc_vals);
+      LIBMESH_CHKERR(ierr);
+      // clear data for this row
+      sub_cols.clear();
+      sub_vals.clear();
+    }
+  }
+  ierr = MatAssemblyBegin(petsc_submatrix->_mat,MAT_FINAL_ASSEMBLY);
+  LIBMESH_CHKERR(ierr);
+  ierr = MatAssemblyEnd(petsc_submatrix->_mat,MAT_FINAL_ASSEMBLY);
+  LIBMESH_CHKERR(ierr);
```

Comment:
I think these 4 lines can be replaced by the recently-added macro:
```
MatAssemblyBeginEnd(petsc_submatrix->comm(), petsc_submatrix->_mat, MAT_FINAL_ASSEMBLY);
```

Reply:
That is right. I will make the change. Thanks!